### PR TITLE
feat(client): enforce `driverAdapters` preview feature

### PR DESCRIPTION
**Pull request by aqrln**
* Enforce `driverAdapters` preview feature to use driver adapters at run time and not just on type level.
* Minor refactoring along the way: make `getPrismaClient` options validators type-safe.
* Add functional tests helpers for testing infrastructure around driver adapters.

### CODE REVIEW

File: packages/client/tests/functional/_utils/types.ts [hunk 1]
```diff
@@ -5,6 +5,7 @@ export type MatrixOptions = {
     from: `${Providers}`[]
     reason: string
   }
+  skipBinary?: boolean
```
Comment (SevInf) [inline packages/client/tests/functional/_utils/types.ts:8]: Should we require a  reason here as well?

File: packages/client/src/runtime/utils/validatePrismaClientOptions.ts [hunk 2]
```diff
@@ -50,18 +60,26 @@ It should have this form: { url: "CONNECTION_STRING" }`,
       }
     }
   },
-  adapter: (_options: any) => {
-    return
+  adapter: (adapter, config) => {
+    if (adapter === undefined) {
```
Comment (SevInf) [inline packages/client/src/runtime/utils/validatePrismaClientOptions.ts:64]: At some point me and Joël discussed if stricter validation for `adapter` makes things safer:

1. Always throw on explicit `undefined` value, but not on missing property.
2. Allow `null` for explicitly skipping driver adapter.

Rationale for that was that if `adapter` is `undefined` by accident, it many cases it is not really obvious that adapter is not used. We ourselves have been burned by that couple of times. Doing validation this way prevents this mistake, and those who want to use adapter conditionally can still pass `null`.

What would you think about that?

**Review by SevInf — APPROVED** [review]
I left a couple of suggestions. I don't think any of them are blocking and strictly required for this to get merged, feel free to ignore or do them in follow up PRs.

**Review by Druue — APPROVED** [review]

File: packages/client/src/runtime/utils/getPreviewFeatures.ts
Comment (millsp) [inline packages/client/src/runtime/utils/getPreviewFeatures.ts:1]: Could we move this file to `runtime/core/init`?

**Review by millsp — APPROVED** [review]
Neat!

**Comment by millsp** [comment]
The failing tests are expected as we are automatically adding the preview feature flags, that will mess with your test. You could use `skipTestIf(driverAdapter)(...`, this variable is available in `clientMeta`.

File: packages/client/src/runtime/utils/validatePrismaClientOptions.ts (hunk 2 above)
Comment (aqrln) [reply packages/client/src/runtime/utils/validatePrismaClientOptions.ts:64]: Absolutely, will do this in a follow-up PR
https://github.com/prisma/team-orm/issues/388

**Pull request merged**